### PR TITLE
Refactor product listing feature subscriber to have a more service or…

### PR DESCRIPTION
**Pull request by OliverSkroblin**
…iented approach

### 5. Checklist

- [ ] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfil them.

### <samp>🤖 Generated by Copilot at 8eed975</samp>

This pull request introduces a new feature for the product listing domain, which allows displaying and filtering products on the storefront by different criteria. It adds several new classes that implement different listing processors and filter handlers, which are responsible for preparing and processing the product listing criteria and result. It also adds a feature flag for the version 6.6.0.0 that changes the behavior of the `ProductListingRoute` class and deprecates the `ProductListingResultEvent`. It also adds a new exception type and a factory method to `ProductException` for handling invalid sorting keys. It also modifies the dependency injection configuration, the phpstan baseline file, and some existing classes to support the new feature.

### <samp>🤖 Generated by Copilot at 8eed975</samp>

*  Introduce the listing processor and filter handler features for the product listing route, which allow preparing and processing the product listing criteria and result by using different classes that implement a common interface (F3-F16, [link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-67620a40765206835ca0e40c0b04721723d8290e4fd38938322f1ec5f44e5633L6-R9), [link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-67620a40765206835ca0e40c0b04721723d8290e4fd38938322f1ec5f44e5633L22-R28), [link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-67620a40765206835ca0e40c0b04721723d8290e4fd38938322f1ec5f44e5633L34-R60))
* Add a feature flag for the version 6.6.0.0, which is used to conditionally execute the new or old code for the product listing route, and deprecate the event dispatching mechanism that was used to customize the product listing criteria and result ([link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-33ce2e7b9d71b434ea7fdbe67ee6ee707266923e27b318a7f14e37e6e7237b49R15), [link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-33ce2e7b9d71b434ea7fdbe67ee6ee707266923e27b318a7f14e37e6e7237b49L70-R75), [link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-67620a40765206835ca0e40c0b04721723d8290e4fd38938322f1ec5f44e5633L34-R60))
* Remove an error message from the `phpstan-v66-baseline.neon` file, which was related to a code style rule that forbids throwing new exceptions within classes ([link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-3254a39100645458b1d776980249db39dd5e8db6d33cdcc15803bb68a128be03L881-L885))
* Add several new service definitions to the `product.xml` file, which are used to configure the dependency injection container for the product domain, and register the new listing processor and filter handler classes with the appropriate tags ([link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-817ef170e1c15f14fffba40fe7f1959831b7121c654f67f77ade44630788e982L235-R285))
* Add a new constant and a new static method to the `ProductException` class, which are used to define and create a new exception type for the sorting feature ([link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-7637f147e1ea4fc286700820ed72b9bc94798160bb05fbf8d5f344973b4a2cb8R16), [link](https://github.com/shopware/platform/pull/3085/files?diff=unified&w=0#diff-7637f147e1ea4fc286700820ed72b9bc94798160bb05fbf8d5f344973b4a2cb8R28-R37))

### CODE REVIEW

File: src/Core/Content/Product/SalesChannel/Listing/Processor/EmptyProcessTrait.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+use Shopware\Core\Content\Product\SalesChannel\Listing\ProductListingResult;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Symfony\Component\HttpFoundation\Request;
+
+#[Package('inventory')]
+trait EmptyProcessTrait
+{
+    public function process(Request $request, ProductListingResult $result, SalesChannelContext $context): void
```
Comment (keulinho) [inline src/Core/Content/Product/SalesChannel/Listing/Processor/EmptyProcessTrait.php]: do we  need this trait? do we actually want this?

For me this is magic which just hides information, why is this a trait at all? 
Can't we implement this method in the abstract classes as empty and not declare it as abstract when in fact it is optional?

That also solves the issue that currently we use one trait for two interfaces, which means we can't change the interfaces separately anymore

File: src/Core/Content/Product/SalesChannel/Listing/Filter/AbstractFilterHandler.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    abstract public function create(Request $request, SalesChannelContext $context): ?Filter;
+
+    /**
+     * The `process` function allows the developer, to post-process the calculated listing result and further process
+     * the determined aggregation values to a more user readable state.
+     *
+     * If you don't have a process use case, you can simply use the `EmptyProcessTrait.php` to fulfill the interface.
+     */
+    abstract public function process(Request $request, ProductListingResult $result, SalesChannelContext $context): void;
```
Comment (keulinho) [inline src/Core/Content/Product/SalesChannel/Listing/Filter/AbstractFilterHandler.php]: don't mark it as abstract and add an empty base implementation if it is optional, same with the other abstract class and remove the trait, see comment on the trait as well

File: src/Core/Content/Product/SalesChannel/Listing/Filter/AbstractFilterHandler.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+     * How the filter should be added to the criteria, is handled by the `AggregationProcessor`, which handles
+     * the global behavior, how the filters should be calculated.
+     */
+    abstract public function create(Request $request, SalesChannelContext $context): ?Filter;
+
+    /**
+     * The `process` function allows the developer, to post-process the calculated listing result and further process
+     * the determined aggregation values to a more user readable state.
+     *
+     * If you don't have a process use case, you can simply use the `EmptyProcessTrait.php` to fulfill the interface.
```
Comment (mitelg) [inline src/Core/Content/Product/SalesChannel/Listing/Filter/AbstractFilterHandler.php]: this sentence is now obsolete

**Comment by keulinho** [comment]
Thanks for your contribution, your PR was merged!

**Pull request closed without merging**